File: _raw/example1/src/main/java/com/example/grpc/echo/unary/UnaryEchoFutureClient.java
package com.example.grpc.echo.unary;

import com.google.common.util.concurrent.FutureCallback;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.MoreExecutors;
import io.grpc.Grpc;
import io.grpc.InsecureChannelCredentials;
import io.grpc.ManagedChannel;
import io.grpc.Status;
import com.example.grpc.echo.EchoRequest;
import com.example.grpc.echo.EchoResponse;
import com.example.grpc.echo.EchoServiceGrpc;
import com.example.grpc.echo.Logging;

import java.util.concurrent.TimeUnit;
import java.util.logging.Logger;

public class UnaryEchoFutureClient {

    private static final Logger logger = Logger.getLogger(UnaryEchoFutureClient.class.getName());

    public static void main(String[] args) throws Exception {
        Logging.init();

        ManagedChannel channel = Grpc.newChannelBuilder("localhost:50051", InsecureChannelCredentials.create()).build();

        EchoServiceGrpc.EchoServiceFutureStub futureStub = EchoServiceGrpc.newFutureStub(channel);
        ListenableFuture<EchoResponse> responseFuture = futureStub.unaryEcho(EchoRequest.newBuilder().setMessage("world").build());
        Futures.addCallback(responseFuture, new FutureCallback<EchoResponse>() {
            @Override
            public void onSuccess(EchoResponse response) {
                logger.info("response received: " + response.getMessage());
            }

            @Override
            public void onFailure(Throwable t) {
                logger.warning("error: " + Status.fromThrowable(t));
            }
        }, MoreExecutors.directExecutor());

        channel.shutdown().awaitTermination(10, TimeUnit.SECONDS);
    }
}

File: _raw/kotlin/src/main/java/com/example/grpc/echo/unary/UnaryEchoFutureClient.kt
package com.example.grpc.echo.unary

import com.example.grpc.echo.EchoRequest
import com.example.grpc.echo.EchoResponse
import com.example.grpc.echo.EchoServiceGrpc
import com.example.grpc.echo.Logging
import com.google.common.util.concurrent.FutureCallback
import com.google.common.util.concurrent.Futures
import com.google.common.util.concurrent.MoreExecutors
import io.grpc.Grpc
import io.grpc.InsecureChannelCredentials
import io.grpc.Status
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.logging.Logger

object UnaryEchoFutureClient {
    private val logger: Logger = Logger.getLogger(UnaryEchoFutureClient::class.java.getName())

    @kotlin.Throws(Exception::class)
    @kotlin.jvm.JvmStatic
    fun main(args: Array<String>) {
        Logging.init()

        val channel = Grpc.newChannelBuilder("localhost:50051", InsecureChannelCredentials.create()).build()

        val futureStub = EchoServiceGrpc.newFutureStub(channel)
        val request = EchoRequest.newBuilder().setMessage("world").build()
        val responseFuture = futureStub.unaryEcho(request)

        val latch = CountDownLatch(1)
        Futures.addCallback<EchoResponse?>(responseFuture, object : FutureCallback<EchoResponse?> {
            override fun onSuccess(response: EchoResponse) {
                logger.info("result: " + response.getMessage())
                latch.countDown()
            }

            override fun onFailure(t: Throwable) {
                logger.warning("error: " + Status.fromThrowable(t))
                latch.countDown()
            }
        }, MoreExecutors.directExecutor())

        latch.await()
        channel.shutdown().awaitTermination(10, TimeUnit.SECONDS)
    }
}

